drivers/random/random_stm32_rng.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2017 Erwin Rol <erwin@erwinrol.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <random.h>
+#include <drivers/rand32.h>
+#include <init.h>
+#include <misc/__assert.h>
+#include <misc/util.h>
+#include <errno.h>
+#include <soc.h>
+#include <misc/printk.h>
+#include <clock_control.h>
+#include <clock_control/stm32_clock_control.h>
+
+#include <stm32f4xx_hal_rng.h>
+
+#include "random_stm32_rng_priv.h"
+
+static int random_stm32_rng_get_entropy(struct device *dev, u8_t *buffer,
+					u16_t length)
+{
+	struct random_stm32_rng_dev_data *dev_data;
+	u32_t rndbits;
+
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(buffer != NULL);
+
+	dev_data = DEV_DATA(dev);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	while (length > 0) {
+
+		if (HAL_RNG_GenerateRandomNumber(&(dev_data->hrng), &rndbits) != HAL_OK) {
+			return -EIO;
+		}
+
+		switch (length) {
+		default:
+			*buffer = rndbits & 0xFF;
+			buffer++;
+			length--;
+			rndbits >>= 8;
+			/* fall through */
+		case 3:
+			*buffer = rndbits & 0xFF;
+			buffer++;
+			length--;
+			rndbits >>= 8;
+			/* fall through */
+		case 2:
+			*buffer = rndbits & 0xFF;
+			buffer++;
+			length--;
+			rndbits >>= 8;
+			/* fall through */
+		case 1:
+			*buffer = rndbits & 0xFF;
+			buffer++;
+			length--;
+			break;
+		}
+	}
+
+	return 0;
+}
+
+static inline void __random_stm32_rng_get_clock(struct device *dev)
+{
+	struct random_stm32_rng_dev_data *dev_data = DEV_DATA(dev);
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	__ASSERT_NO_MSG(clk);
+
+	dev_data->clock = clk;
+}
+
+static int random_stm32_rng_init(struct device *dev)
+{
+	struct random_stm32_rng_dev_data *dev_data;
+	struct random_stm32_rng_dev_cfg *dev_cfg;
+
+	__ASSERT_NO_MSG(dev != NULL);
+
+	dev_data = DEV_DATA(dev);
+	dev_cfg = DEV_CFG(dev);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+	__ASSERT_NO_MSG(dev_cfg != NULL);
+
+	__random_stm32_rng_get_clock(dev);
```

Comment:
Just used other stm32 drivers like pwm_stm32 and uart_stm32 as example, but I'll change it to inline.